### PR TITLE
Add CTC align op

### PR DESCRIPTION
**Pull request by wanghaoshuang**

### CODE REVIEW

File: paddle/operators/ctc_greedy_decode_op.cu [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    const T* logits = input->data<T>();
+    Tensor tmp;
+    int* tokens = tmp.mutable_data<int>({num_tokens, 1}, ctx.GetPlace());
+    // get argmax
+    // platform::GpuMemsetAsync(args, 0, sizeof(float), stream);
+
+    auto stream = ctx.cuda_device_context().stream();
+    ArgmaxCudaKernel<T, PADDLE_CUDA_NUM_THREADS><<<
+        num_tokens, PADDLE_CUDA_NUM_THREADS, 0, stream>>>(seq_width, logits,
+                                                          tokens);
```
Comment (qingqing01) [inline paddle/operators/ctc_greedy_decode_op.cu]: 这个Kernel是在计算top 1吗？如果是可以调用top_k_op的实现吧~

File: paddle/operators/ctc_greedy_decode_op.cc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+class CTCGreedyDecodeOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  CTCGreedyDecodeOpMaker(OpProto* proto, OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Input",
+             "(LodTensor, default: LoDTensor<float>), the unscaled "
+             "probabilities of variable-length sequences, which is a 2-D "
+             "Tensor with LoD information. It's shape is "
```
Comment (kuke) [inline paddle/operators/ctc_greedy_decode_op.cc]: It's -> Its

**Comment by qingqing01** [comment]
Please create an issue and add it to https://github.com/PaddlePaddle/Paddle/projects/39

File: python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from op_test import OpTest
+from test_softmax_op import stable_softmax
+
+
+def CTCGreedyDecode(softmax, blank, merge_repeated):
+    prev_token = -1
+    result = []
+    for token in np.argmax(softmax, axis=1):
+        if (token != blank) and not (merge_repeated and token == prev_token):
+            result.append(token)
```
Comment (kuke) [inline python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py]: Should there be one line `prev_token = token`?

File: python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        self.outputs = {"Output": output}
+        self.attrs = {
+            "blank": self.blank,
+            "merge_repeated": self.merge_repeated
+        }
+
+    def test_check_output(self):
+        self.check_output()
+
+
```
Comment (kuke) [inline python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py]: Please add another test case for `merge_repeated = False`

File: paddle/operators/ctc_greedy_decode_op.cc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        ctx.device_context());
+  }
+};
+
+class CTCGreedyDecodeOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  CTCGreedyDecodeOpMaker(OpProto* proto, OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Input",
+             "(LodTensor, default: LoDTensor<float>), the unscaled "
```
Comment (kuke) [inline paddle/operators/ctc_greedy_decode_op.cc]: unscaled -> unnormalized

File: paddle/operators/ctc_greedy_decode_op.cc [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                 "(int, default: 0), the blank label setted in Connectionist "
+                 "Temporal Classification (CTC) op, and it is in the "
+                 "half-opened interval [0, num_classes + 1).")
+        .SetDefault(0);
+    AddAttr<bool>("merge_repeated",
+                  "(bool, default: true), whether to "
+                  "merge repeated elements between two blanks. ")
+        .SetDefault(true);
+    AddComment(R"DOC(
+CTCGreedyDecoder is an implementation of the simple best path decoding
```
Comment (kuke) [inline paddle/operators/ctc_greedy_decode_op.cc]: Need more detailed document here

File: python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py (hunk 4 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py]: Thx. Fixed.

File: python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py (hunk 3 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py]: Thx. Fixed.

File: python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py (hunk 4 above)
Comment (wanghaoshuang) [reply python/paddle/v2/fluid/tests/test_ctc_greedy_decode.py]: Thx. Fixed.

File: paddle/operators/ctc_greedy_decode_op.cu (hunk 1 above)
Comment (wanghaoshuang) [reply paddle/operators/ctc_greedy_decode_op.cu]: You are right. I will remove  `argmax` content from both `CPU` kernel and `GPU` kernel.

File: paddle/operators/ctc_greedy_decode_op.cc (hunk 5 above)
Comment (wanghaoshuang) [reply paddle/operators/ctc_greedy_decode_op.cc]: Done.

File: paddle/operators/ctc_greedy_decode_op.cc (hunk 2 above)
Comment (wanghaoshuang) [reply paddle/operators/ctc_greedy_decode_op.cc]: Thx. Done.

File: paddle/operators/ctc_greedy_decode_op.cc (hunk 6 above)
Comment (wanghaoshuang) [reply paddle/operators/ctc_greedy_decode_op.cc]: Thx. Done.

**Comment by kuke** [comment]
Please keep the name `CTCGreedyDecode`, we would add beam search decoding soon.

File: paddle/operators/ctc_decode_op.cu [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    thrust::device_vector<size_t> dev_out_lod0(input_lod[level].size());
+    size_t* dev_out_lod0_ptr = thrust::raw_pointer_cast(dev_out_lod0.data());
+
+    // merge elements and delete blank
+    T* output_data = output->mutable_data<T>({num_tokens, 1}, ctx.GetPlace());
+
+    auto stream = ctx.cuda_device_context().stream();
+    MergeAndDelCudaKernel<T><<<1, 1, 0, stream>>>(
+        num_tokens, tokens, num_seq, input_lod[level].data(), blank,
+        merge_repeated, dev_out_lod0_ptr, output_data);
```
Comment (qingqing01) [inline paddle/operators/ctc_decode_op.cu]: The CUDA kernel is less efficient here. We can profile the speed when training the model. Then determine whether to delete the GPU kernel in this op and editing distance op.

**Comment by wanghaoshuang** [comment]
Have removed debug code.

**Pull request merged**